src/Main.elm

```diff
@@ -1,10 +1,571 @@
 module Main exposing (main)
 
-import Data.Audience
-import Data.AudienceFolder
-import Html exposing (Html)
+import Browser
+import Data.Audience exposing (Audience, AudienceType(..), audiencesJSON)
+import Data.AudienceFolder exposing (AudienceFolder, audienceFoldersJSON)
+import Dict exposing (Dict)
+import FontAwesome as FA
+import Html exposing (..)
+import Html.Attributes exposing (class, classList, id)
+import Html.Events exposing (onClick)
+import Json.Decode as Decode exposing (Decoder, decodeString, field, float, int, list, nullable, string)
+import Json.Decode.Pipeline exposing (optional, required, requiredAt)
+import Tree as Tree exposing (Tree)
+import Tree.Zipper as Zipper
 
 
-main : Html msg
+type Node
+    = Dir Id AudienceFolder
+    | File Id Audience
+    | Root
+
+
+type alias Id =
+    Int
+
+
+
+{--------------------------------------------------------------
+
+This program mimics directory structure where there are files and folders.
+Files are Audience and Folders are AudienceFolder. 
+Once decoded from JSON, they are converted to List Node
+
+List Node is then ordered in a Rose Tree using zwilias/elm-rosetree/1.5.0 library.
+
+At first all nodes are entered into rose tree with flat architecture, see flatTree
+Parent child relation is ignored.
+The reason for this is, you cannot put a node into its parent when the parent itself is not yet traversed and is not in the tree yet!
+Once you have all the nodes in a flat tree and information about { parent : Maybe Node, child : Node }, then 
+1. you can pluck a node from the tree
+2. find its parent in the tree
+3. put it inside the parent, outputting updated tree
```

Comment:
Here's a magic trick: it is possible to create a tree using this parent-child information. There's a nice function for it: [`Tree.unfold`](https://package.elm-lang.org/packages/zwilias/elm-rosetree/latest/Tree#unfold)

Example usage: https://ellie-app.com/fgPcP2M5wPda1

I think it would remove the need for all these functions:
* `flatTreeToParentChildTree`
* `reorderNode`
* `appendASubTreeAsChild`
* `findASubTree`
* `removeASubTree`
* and probably more

Anyways, it's impressive that you were able to do things explicitly, by hand... :)

Reply:
Thanks for the example. I have noticed `Tree.unfold` but didn't quite get it at the time.